### ISSUE TITLE
support control flow analysis of tagged template calls

### ISSUE THREAD

**Issue by turadg** [microsoft/TypeScript#51426]
# Suggestion

## 🔍 Search Terms

return never tagged template unreachable code

Related issues:
https://github.com/microsoft/TypeScript/pull/32695
https://github.com/microsoft/TypeScript/issues/50363

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Control flow analysis detecting unreachable code after a tagged template function call that throws.

## 📃 Motivating Example

It's a surprise that the same function call does not have control flow analysis when used as a tagged template:
```
function throws(opt?: any): never {
    throw new Error()
}

function testFn() {
    throws(`reason`);
    
    // @ts-expect-error unreachable
    console.log("is unreachable!")
}

function testTag() {
    throws`reason`;
    
    // ts does not detect that this is unreachable
    console.log("is unreachable!")
}
```
([playground](https://www.typescriptlang.org/play?target=99&noFallthroughCasesInSwitch=true&ts=4.8.4#code/GYVwdgxgLglg9mABFAFgJzgdwM4Ao4AOUA-AFyICGYAngJTlgCmAbo2ogN4BQivy6WRE0yIAomgxpctLgF8uXUJFgJkjbFABiYaZx59UGHLgAGaRhWwITtANz7eDxAHpniAAJRsAWkYAPAkZoXwk4dnBzCggUCgAjABtGJwgEK0SAOni4AHNcACIYbEQIi2i4xIBCPJl5RXBoeCQodSgAFQpc2j0+fiNsMwsrMBN7HqdXZCKAEzh1ITgoRCnGZuh+CkXUQsRtkqiYhKSelLA0xkyc-N2wSLLDqpquIA))

### Throwing an exception with a message

Normally this would be something like,
```js
assert(condition, 'explanation');
```
However for performance reasons we eschew the function call:
```
condition || Fail`explanation`;
```

The `asserts` works but the `Fail` doesn't. We could work around this by `Fail('explanation')` but we prefer to be consistent in use of the `Fail` function. So for our current work-around is to explicitly `throw` the tagged template.

**Comment by fatcerberus** [comment]
This is actually two separate issues.  The pattern `condition || fail("explanation")` doesn't work anyway, as only top-level function calls affect control flow analysis.

https://github.com/microsoft/TypeScript/pull/32695

> A function call is analyzed as an assertion call or never-returning call when ...
> ... the call occurs as a top-level expression statement

**Comment by MartinJohns** [comment]
> Do `switch` case blocks count as a top-level expression statement?

A `switch` is not an expression at all.

**Comment by turadg** [comment]
For `condition || fail("explanation") ` @mhofman filed https://github.com/microsoft/TypeScript/issues/50739

**Comment by mhofman** [comment]
The lack of narrowing based on logical expression was previously filed as issue #50739.

We can focus this issue on the lack of control flow analysis through template literal as the following code would show:


```ts
function throws(opt?: any): never {
    throw new Error()
}

declare const a: string | null;

function testFn() {
    if (!a) throws(`reason`);
    
    a.charAt(0); // a is correctly narrowed to `string`
}

function testTag() {
    if (!a) throws`reason`;
    
    a.charAt(0); // Object is possibly 'null'. ts(2531)
}

function testFnInSwitch() {
    switch(a) {
        case null: // No fallthrough correctly infered
            throws(`reason`);
        default:
            a.charAt(0); // a is correctly narrowed to `string`
    }
}

function testTagInSwitch() {
    switch(a) {
        case null: // Fallthrough case in switch. ts(7029)
            throws`reason`;
        default:
            a.charAt(0);  // Object is possibly 'null'. ts(2531)
    }
}
```

**Playground Link:** [Provided](https://www.typescriptlang.org/play?target=99&noFallthroughCasesInSwitch=true&ts=4.8.4#code/GYVwdgxgLglg9mABFAFgJzgdwM4Ao4AOUA-AFyICGYAngJTlgCmAbo2ogN4BQivy6WRE0yIAomgxpctLgF8uXACaMIAGwppGiCAmxRK5PWhhgA5ogA+QkKtUBuBaEiwEyRnoBiYaZx58YwIi4AIQUtPwYOLgABpoU2AjRtA58iH68FAB0ECgaAIJQuAAMyYgA9GWUiDDY2nASKlCq1EIakYyKyHCI0UYmptFyjuDQ8EhQ7lAAKhSmPtypAUGh4aiR2LGM8YkpfOmU2bloBcWlFYgA8gBGAFaN1bUEcNjYMFfNiADkYDaqn5nIPAAJgArABmACMMnkXCco1cE08YAAkmAAMqYGBQHLzfbYTHYlC4MK+VKpCDxLQ-Wzkc4AOW6wAotjWcBAphQdQa0A+JmAbA6+zJfFZUU22zASV2wuUTJsUFIQuFBxy+UKJTs5UqFAeXM0PJaYDaWA6XR6fTMg1S8hhcJc40mM1MqIxWJx4QWfHxbqJJM9wop2Cpv1plQ8zNUrPZnMDWhMiG9hIBUDwAHYikCAJwyZVk0UbOIJSXSsmyijyxW51JZVXHdWlLWXW73GqIJ4vN4fb6-f6A3CgyE5vg2oA)

**Comment by 71** [comment]
I have a prototype implementation in https://github.com/71/TypeScript/commit/de96190401979ebebbc7089fe5ac87673f1a6191, but can't submit a PR as this issue is not in the "Backlog" milestone. Is there any way this could be added to the backlog? The proposed change is fairly minimal, but fixes this issue.

**Comment by 71** [comment]
Gentle ping (sorry) -- would love to submit a PR for this.

**Comment by GongT** [comment]
[playground](https://www.typescriptlang.org/play/?ts=5.9.0-dev.20250604#code/GYVwdgxgLglg9mABMAhlFAbAFABxQJygGcAuRIqfGMAc0QB9EAVAUwFscM0WBlS6mkQCC+fCgCeAGkQA6OQUFkUYcQG0AugEoyYFgDcW+RAG8AUIkQQEROBhYyMcGlgBEqdBkuY7AE0QB3GCgACxIXaTxCImkFIk0AbnNEAHpkxBx8OAgWIiIZFgAPIKwARgSkkMz-RF1qgFFROHxXGzYWEIEAzNoAQhdygF9TU1BIWAREKByoISxNEySrMApEFDIKKloGRHAfFmBqFj8AXkQABlWiVZVEixhgRCwelHmzCwt3TAADYJYMR0QABJjCUAEwAZgGPS+t0QQwsSxsdgcTlcKGO4XSBCILAAkmAoFgXoNhqNoPAkFMKAAhOYLBHWKCIABG634W0Yu32hxO50u13EsPujx6zNeSQ+aEwWB+fwBwLBkOh5Qs8Ms1ls9kczhczIxEWxeIJWDFJNMVJmc3iQA)

So anyone add this to backlog?🥺
